contracts/Staking.sol

```diff
@@ -0,0 +1,488 @@
+    uint256 convertedAmount = amount.div(TEN_DECIMALS);
+    uint256 _oracleRelayerFee = (msg.value).sub(amount);
+
+    bytes[] memory elements = new bytes[](4);
+    elements[0] = EVENT_DELEGATE.encodeUint();
+    elements[1] = msg.sender.encodeAddress();
+    elements[2] = validator.encodeAddress();
+    elements[3] = convertedAmount.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
+    ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(CROSS_STAKE_CHANNELID, _RLPEncode(EVENT_DELEGATE, msgBytes), _oracleRelayerFee.div(TEN_DECIMALS));
+    payable(TOKEN_HUB_ADDR).transfer(msg.value);
+
+    delegated[msg.sender] = delegated[msg.sender].add(amount);
+    delegatedOfValidator[msg.sender][validator] = delegatedOfValidator[msg.sender][validator].add(amount);
+
+    emit delegateSubmitted(msg.sender, validator, amount, _oracleRelayerFee);
+  }
+
+  function undelegate(address validator, uint256 amount) override external payable tenDecimalPrecision(amount) initParams {
+    require(pendingUndelegated[msg.sender][validator] == 0, "pending undelegation exist");
+    amount = amount != 0 ? amount : delegatedOfValidator[msg.sender][validator];
+    if (amount < minDelegationChange) {
+      require(amount == delegatedOfValidator[msg.sender][validator],
+        "the amount must not be less than minDelegationChange, or else equal to the remaining delegation");
+    }
+    delegatedOfValidator[msg.sender][validator] = delegatedOfValidator[msg.sender][validator].sub(amount, "not enough funds to undelegate");
+
+    // native bnb decimals is 8 on BBC, while the native bnb decimals on BSC is 18
+    uint256 convertedAmount = amount.div(TEN_DECIMALS);
+    uint256 _oracleRelayerFee = msg.value;
+    require(_oracleRelayerFee >= oracleRelayerFee, "the msg value should be no less than the minimum oracleRelayerFee");
+
+    bytes[] memory elements = new bytes[](4);
+    elements[0] = EVENT_UNDELEGATE.encodeUint();
+    elements[1] = msg.sender.encodeAddress();
+    elements[2] = validator.encodeAddress();
+    elements[3] = convertedAmount.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
+    ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(CROSS_STAKE_CHANNELID, _RLPEncode(EVENT_UNDELEGATE, msgBytes), _oracleRelayerFee.div(TEN_DECIMALS));
+    payable(TOKEN_HUB_ADDR).transfer(_oracleRelayerFee);
+
+    delegated[msg.sender] = delegated[msg.sender].sub(amount);
+    pendingUndelegated[msg.sender][validator] = amount;
+
+    emit undelegateSubmitted(msg.sender, validator, amount, _oracleRelayerFee);
+  }
+
+  function redelegate(address validatorSrc, address validatorDst, uint256 amount) override external payable tenDecimalPrecision(amount) initParams {
+    require(validatorSrc!=validatorDst, "invalid redelegation");
+    amount = amount != 0 ? amount : delegatedOfValidator[msg.sender][validatorSrc];
+    if (amount < minDelegationChange) {
+      require(amount == delegatedOfValidator[msg.sender][validatorSrc],
```

Comment:
I would better like that when the delegatedOfValidator[msg.sender][validatorSrc] is little than minDelegationChange, the reldelegate action should be forbidden.

Reply:
If so, when the balance is less than minDelegationChange, the delegator must undelegate first to get his balance back

Reply:
I think so.